[SNAP-2121] separate diskstore for metastore and delta regions

sumwale opened the pull request:
Use separate diskstores for metastore and delta regions. Datadictionary one for former and
new one for latter. However, backward compatibility has to be also maintained so the diskstore
assignment is done by modifying the DDLs that get executed (latter from Snappy layer) so
when starting from existing cluster where these tables have already been created, it will
behave as before.

## Changes proposed in this pull request

- set the "datadictionary" to be the diskstore when the connection specifies default schema as the
  SNAPPY_HIVE_METASTORE which should only be done by the internal hive metastore client
- make all the regions using "datadictionary" as diskstore as persistent in DistributionDefinitionNode
  (in particular metastore regions will be persistent on locator now so backing up datadictionary
   will backup full metadata)
- added separate diskstore in "snappy-internal-delta" for delta regions
- make entry for it in SYSDISKSTORES (in FabricDatabase bootup)
- updated tests for the same

## Patch testing

precheckin

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/916

Commit:
[SNAP-2121] separate diskstore for metastore and delta regions

- set the "datadictionary" to be the diskstore when the connection specifies default schema as the
  SNAPPY_HIVE_METASTORE which should only be done by the internal hive metastore client
- make all the regions using "datadictionary" as diskstore as persistent in DistributionDefinitionNode
  (in particular metastore regions will be persistent on locator now so backing up datadictionary
   will backup full metadata)
- added separate diskstore in "snappy-internal-delta" for delta regions
- make entry for it in SYSDISKSTORES (in FabricDatabase bootup)
- add a flag in DDLConflatable to indicate whether hive-metastore should be in datadictionary or not;
  this is for backward compatibility with previous releases where it was in default diskstore that
  should continue to be honoured (due to absence of the new flag in those persisted DDLs)
- updated tests for above changes

kneeraj commented:
@sumwale - So even if the user has created a diskstore of his own and given that disk store for persistence of the column table the row buffer will always go in this internal delta diskstore?

Commit:
formatting changes

Commit:
use common method in GemFireStore.createDiskStore

this method has retries to deal with transient disk exceptions on NFS mounts etc
so use it consistently for creating all internal diskstores (including datadictionary)

sumwale commented:
@kneeraj Yes, otherwise compactor keeps copying tons of column data unnecessarily.

kneeraj commented:
Another approach could have been to just modify the needsCompaction() method to return true or false based on the size which is valid instead of the number of live entries. Actually this should be the case anyways for even normal tables where rowSizes vary a lot. Plus all sort of size accounting is already there.
I mean disk capacity planning needs to be more involved for the user now. If there are several column tables then this internal diskstore also need to be of a pretty decent size.

kneeraj commented:
In that approach the backward compatibility concern is also not there.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -1199,30 +1197,26 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
+        long start = 0;
+        if (logger.infoEnabled()) {
+          start = System.currentTimeMillis();
+          logger.info("FabricDatabase: waiting for index loading from "
```
Now I feel that these log lines should be at the config level also. If you agree then please do the change accordingly.

sumwale replied:
config is < info, so all info logs do get logged at config level too.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -1453,17 +1447,15 @@ private void clearAllIndexes(ArrayList<GemFireContainer> uninitializedContainers
-        logger.info("FabricDatabase: recreateAllLocalIndexes " +
-            "index re-creation for disk store: " + ds.getName() + " ended");
-      }
+      PersistentOplogSet oplogSet = ds.getPersistentOplogSet(null);
```
Ok. So since you are using GFXD_DD_DISKSTORE_NAME for Hive metatables as well so for their indexes you removed this condition?
 if (!ds.getName().equals(GfxdConstants.GFXD_DD_DISKSTORE_NAME))

sumwale replied:
Yes hive metatables create indexes. These are still excluded from "persist-indexes" since the datadictionary diskstore is marked "isForInternalUse" but index recreation including consistency check should be done for that to be on the safe side (and recreate index if mismatch).

sumwale commented:
@kneeraj No cannot use that approach else the krf sizes will keep increasing. This will cause restart to go OOM as reported on Slack (compactor not compacting for that case is another issue but we need compaction else it will cause lot of trouble later).

sumwale commented:
Also a general change to compactor to be based on sizes rather than count can also cause trouble otherwise for similar reasons because lots of redundant keys can build up which can cause problems in recovery map creation on restart.

About disk size planning, yes we may need to document if user is doing lots of jdbc inserts. Don't think the size can exceed 1-2GB even for multiple column tables but we can document for the case when non-default diskstore has been configured. Are we documenting disk sizes for default diskstore when no explicit diskstore is specified? Normally its understood that user has lots of disk space. Planning there is very difficult in any case due to the way compactor runs, so we can't really be sure that even 4X disk space would be enough though normally that is the recommendation.

kneeraj commented:
>> Also a general change to compactor to be based on sizes rather than count can also cause trouble otherwise for similar reasons because lots of redundant keys can build up which can cause problems in recovery map creation on restart.

No, this should not happen because recovery map is built only after all the drfs are read. So all the deleted keys ( delta buffer deleted keys ) will be ignored for recovery map.

----------------------------------------------------------------------------------------------------------------------------------------
I still need to understand this part ( pasting from our offline discussion )


**Neeraj**: When we create the cached batch we anyways go through the entire list of region entries from the row buffer. In that iteration itself if you, using the diskentryId for each entry, you get to know which oplog contains the last valid value for that region entry. Now we can just find out the DiskInfoRegion object corresponding to the rowbuffer region and do the atomic decrement in the info of that region and also decrement the total live count of that oplog. Rest all will automatically happen.

**Sumedh**: As we discussed, there are workarounds possible for deletes but this is a general problem. Column table deletes/updates will suffer the same problem because all of those are marked as updates in crf.

**Neeraj**: In fact the destroy of the entry from rowbuffer should already be doing this. Isn't it?

**Sumedh**: Where?

**Neeraj**: So for some reason the destroy from the rowbuffer for each region entry is not updating the DiskId( oplog id + oplogentryId ) correctly? May be if there is some problem here we just need to fix this. What am I missing?

**Sumedh**: Didn't understand. By this way atmost the last updated Oplog entry will be marked but we need this for all the updates. Perhaps will be fine for rollover of row buffer.Ok, just looked at the code. It does seem to be updating live entry list of previous Oplog after each update/delete in Oplog already. However, compactor is certainly is not working as it ought to. Noticed that sometimes compactor is able to compact all after a 10mins gap or something. Maybe there is some scheduling problem in the code.

So, if the compaction bug which you have pointed is the culprit, may be we just need to fix that. An f2f discussion would help I guess...

-----------------------------------------------------------------------------------------------------------------------------------------------

I am going ahead with the review anyways though. Should be completing soon.

About the docs part, I don't think we do a good job now. But will have to do as per the final design. Will file a task for that.

sumwale commented:
@kneeraj

> No, this should not happen because recovery map is built only after all the drfs are read. So all the deleted keys ( delta buffer deleted keys ) will be ignored for recovery map.

How is that possible? If the drf entry for a key in a separate Oplog how can the recovery map be built after reading that drf first? For each Oplog it has to build the recovery map either way to lookup. Later when drf is read its possible to clear it but temporarily it will take huge memory causing OOM. Also reported on Slack (which would be the obvious case to me).

> So, if the compaction bug which you have pointed is the culprit, may be we just need to fix that. 

Didn't understand. Compator bug should be fixed all right, but we NEED the compaction to happen either way. What exactly is your objection to this PR?

sumwale commented:
@kneeraj 

> So all the deleted keys ( delta buffer deleted keys ) will be ignored for recovery map.

Ok, I see. What you are saying is that a map of all drfs is created first, right? In that case that will cause the OOM. Overhead of krf map and drf map should be similar, and even is latter is somewhat less, the ever increasing number of keys means that restart will likely fail with OOM if compactor has not done the compaction.

kneeraj commented:
>> What you are saying is that a map of all drfs is created first, right?

Yes, it is a Set I guess. If there are a lot of keys then yes there are chances of OOM. This should be considered by persistent design anyways and probably a separate task?

 >> What exactly is your objection to this PR?

Few things which are making me uncomfortable are:

- Data of the same table are in different disk stores. May be user wants to have better disks for table data. [ May be we can create an internal disk-store within the main disk store which user has provided, **or is it already like that**? For example if the DS path is SNAPPY_DELTA_DISKSTORE_NAME then the internal can be  SNAPPY_DELTA_DISKSTORE_NAME/internal ]

- As mentioned above, more information has to be conveyed to the user in docs for provisioning.

The change related to "datadictionary" and hive metastore stuff looks good and logical to me. For separate disk store part I was just trying to see if these can be avoided if alternative exists and that is why discussing few things. As conveyed earlier I am anyways doing the review assuming this design to be valid.

sumwale commented:
> This should be considered by persistent design anyways and probably a separate task?

Yes can be tracked though is low priority with column tables. As had mentioned in the call, could use Spark's sorter (like used in DISTINCT) instead of sets/maps (for both krf and drf) which can overflow to disk if required. With column tables it looks largely unnecessary if we ensure that row buffer keys are properly compacted.

> May be we can create an internal disk-store within the main disk store which user has provided, or is it already like that? For example if the DS path is SNAPPY_DELTA_DISKSTORE_NAME then the internal can be SNAPPY_DELTA_DISKSTORE_NAME/internal

That can be done. My concern there was the doubling of diskstores and thus double the number of compactor threads and other diskstore tasks. Not a problem if those use a common thread pool (perhaps they already do).

kneeraj reviewed (commented):
LGTM

kneeraj commented:
>> That can be done. My concern there was the doubling of diskstores and thus double the number of compactor threads and other diskstore tasks. Not a problem if those use a common thread pool (perhaps they already do).

Well the other problem would be more preblows occupying 1 GB upfront. If user just does the optimized inserts instead of jdbc inserts then very little or no data will come in delta buffer and they will be wasted space.

sumwale commented:
Had reduced the delta diskstore size to 100MB for that, but if there are many of those then diskspace will be wasted. For separate delta diskstores can reduce it further to say 50M.

sumwale commented:
Checked that there are separate thread-pools for each diskstore (some tasks use common waiting pool), but I think having a common pool should be fine and makes sense. Only async writer can have a dedicated pool assuming it to be a rare case. @kneeraj @suranjan what do you think?

kneeraj commented:
Yes each disk store has its own pools. I am fine with a shared one.

sumwale commented:
Have filed a separate task for 1.1 to track separate disk stores for delta regions: https://jira.snappydata.io/browse/SNAP-2127

Commit:
throw proper exception in case metadata fetch fails

instead of failing with NPE throw an IllegalStateException if metadata is unexpectedly null

Commit:
Merge remote-tracking branch 'origin/snappy/master' into SNAP-2121

sumwale commented:
@kneeraj @suranjan 

Followup PRs for SNAP-2127 created for review:
https://github.com/SnappyDataInc/snappy-store/pull/350
https://github.com/SnappyDataInc/snappydata/pull/918
